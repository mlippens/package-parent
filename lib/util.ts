import json5 from 'json5';
import * as path from 'path';
import * as fs from 'fs';

export const loadJSONfile = (filePath: string) => {
    if (!fs.existsSync(filePath)) {
      throw new Error(`Could not find file ${filePath}`);
    }
    return json5.parse(
      fs.readFileSync(filePath, { encoding: "utf-8" })
    );
  };

export const writeJSONfile = (p: string, file: Object) => {
  fs.writeFileSync(
    p,
    JSON.stringify(file, null, 4),
    { encoding: "utf-8"}
  );
}