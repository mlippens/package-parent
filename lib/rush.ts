import * as util from './util';
import * as path from 'path';
import * as fs from 'fs';

type RushConfig = {
    projects: {
      projectFolder: string;
    }[];
  };
  
export const getProjects = async () => {
    const rushJson: RushConfig = util.loadJSONfile('rush.json')
    const rushProjects = await Promise.all(
      rushJson.projects.map(({ projectFolder }) => {
        const p = path.join(__dirname, projectFolder, "package.json");
        if (!fs.existsSync(p)) {
          throw new Error("Could not find package.json on path: " + p);
        }
        return { projectFolder };
      })
    );
    console.debug(`Found ${rushProjects.length} rush projects configured`);
    return rushProjects;
  };
  