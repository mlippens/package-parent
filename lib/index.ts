import * as fs from "fs";
import * as path from "path";
import parseArgs from 'minimist';
import ow from 'ow';
import { merge } from "lodash";
import * as rush from './rush';
import * as util from './util'


type Project = {
  projectFolder: string
}

const getFolders = (p: string) => {
  return fs
    .readdirSync(p, { withFileTypes: true })
    .filter(f => f.isDirectory())
    .map(f => f.name);
}

const exclusions = ['node_modules'];

const findProjects = (p: string) => {
  let currentFolders = getFolders(p).filter(f => !exclusions.includes(f));
  let projects: Project[] = [];
  while (currentFolders.length > 0) {
    const c = currentFolders.pop() as string;
    if (!fs.existsSync(path.join(c, 'package.json'))) {
      getFolders(c)
        .filter(f => !exclusions.includes(f))
        .forEach(f => currentFolders.push(path.join(c, f)));
    } else {
      projects.push({ projectFolder: path.join(p, c) });
    }
  }
  return projects;
}

const configPredicate = ow.object.exactShape({
  packages: ow.array.nonEmpty.ofType(ow.object.exactShape({ 
    folder: ow.string.nonEmpty,
    parent: ow.object.nonEmpty, 
  })),
});


const main = async () => {
  const argv = parseArgs(process.argv.slice(2));
  let projects : Project[] = [];
  if (argv.rush === true) {
    projects = await rush.getProjects();
  } else {
    projects = findProjects(process.cwd());
  }

  if (!argv.c && !argv.config) {
    throw new Error('Expected a -c or --config <config.json> parameter to invoke package-parent!');
  }
  const config = util.loadJSONfile(path.resolve(process.cwd(), argv.c || argv.config));
  ow(config, configPredicate);
  console.info("projects", projects);

  await Promise.all(
    config.packages.map(async (c) => {
      console.info(`processing config=[${c.folder}]`);
      await Promise.all(
        projects
          .filter((p) => p.projectFolder.endsWith(c.folder))
          .map((project) => {
            const packageJsonPath = path.join(project.projectFolder, 'package.json');
            const mergedConfig = merge(util.loadJSONfile(packageJsonPath), c.parent);
            util.writeJSONfile(packageJsonPath, mergedConfig);
            console.debug(
              `Writing package.json of package ${project.projectFolder}`
            );
          })
      );
    })
  );
};

main().catch(console.error);
