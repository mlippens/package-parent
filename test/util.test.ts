import * as fs from 'fs';
import * as path from 'path';
import { loadJSONfile, writeJSONfile } from "../lib/util";

const resolveFile = (f: string) => path.join(__dirname, 'resources', f);

describe('#util', () => {
    describe('#loadJSONfile', () => {
        it('should load an existing json file with comments', () => {
            const file = loadJSONfile(resolveFile('test.jsonc'));
            expect(file).not.toBeNull();
            expect(file).toHaveProperty('answer');
        });
        it('should load an existing json file without comments', () => {
            const file = loadJSONfile(resolveFile('test.json'));
            expect(file).not.toBeNull();
            expect(file).toHaveProperty('answer');
        });
        it('should throw an error when attempting to load a nonexisting file', () => {
            expect(() => loadJSONfile(resolveFile('doesnotexist.json'))).toThrowErrorMatchingSnapshot();
        });
        it('should throw an error when attempting to load a non JSON file', () => {
            expect(() => loadJSONfile(resolveFile('test.txt'))).toThrowErrorMatchingSnapshot();
        }); 
    });
    describe('#writeJSONfile', () => {
        it('should write a json file', () => {
            const file = resolveFile('writetest.json');
            try { 
                const contents = { "answer": 42 };
                expect(() => writeJSONfile(file, contents)).not.toThrowError();
                expect(loadJSONfile(file)).toMatchObject({
                    ...contents,
                })
            } finally {
                fs.rmSync(file);
            }
        });
        it.todo('should fail writing a json file with a non JSON object');
        it.todo('should fail writing a json file to an invalid path');
    });
});