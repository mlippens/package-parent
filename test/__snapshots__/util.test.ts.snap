// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#util #loadJSONfile should throw an error when attempting to load a non JSON file 1`] = `"JSON5: invalid character 'h' at 1:2"`;

exports[`#util #loadJSONfile should throw an error when attempting to load a nonexisting file 1`] = `"Could not find file /home/michael/workspace/package-parent/test/resources/doesnotexist.json"`;
